src/Highway.Data.EntityFramework/ReadonlyDataContext.cs

```diff
@@ -0,0 +1,312 @@
+﻿using System;
```

Comment:
I've been debating if this type should even exist.  Normally, you'd only get an instance through a Repository.  I decided to put it in and override all the write operations from DbContext so you cannot just cast your "sort of readonly thing" to DbContext, Call .Set, and go nuts.  This should actually _enforce_ that you cannot use write operations, whether you create a direct instance or come at it from ReadonlyRepository.

Reply:
I was thinking about this over the weekend, and the context options object could carry the read only configuration in with it, and turn a regular DataContext to readonly without additional objects. Have a gate check for read-only configuration on commit, and keep to the regular context. Seems simpler.

Reply:
@DevlinLiles I don't see any "context options object."

The configuration option is a good one.  My only objection is that any IoC registration now has to be clever enough to resolve that parameter, meaning your IoC provider will need to support some kind of Func registration, and resolve that at runtime based on which Repository you want.

Also, with a single data context (or just no inheritance) we could set ApplicationIntent=ReadOnly on the connection string and provide additional security against writes in IReadonlyDataContext.ExecuteSqlQuery.

I propose:

1. A new overload of the constructors accepting a "readOnly" bool.
2. If "readOnly is true" then:  a)  Set the readonly application intent on the connection string, and b) override all the writing DbContext operations, throwing exceptions if the context is readonly.

The field _readonly should also be readonly, only settable during construction.

Finally, I have to be careful that the public surface of this type expands, but the existing surface doesn't change.  Hence the additional overload vs. any kind of new configuration object.

Are we agreed?

Reply:
yes, more comments below